### PR TITLE
Refactor tool related loco_globals into ToolManager

### PR DESCRIPTION
**Pull request by LeftofZen**
i was unsure whether to put these variables in a separate manager, in Input, or in Ui, as some of them existed all over the place

### CODE REVIEW

File: src/OpenLoco/src/ToolManager.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        return _toolWindowCursor;
+    }
+    void setToolWindowCursor(Ui::CursorId toolWindowCursor)
+    {
+        _toolWindowCursor = toolWindowCursor;
+    }
+
+    // 0x00523394
+    uint16_t getToolWidgetIndex()
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/ToolManager.cpp:45]: Defined as signed on L12 and unsigned here. -1 or 0xFFFF is used to indicate that the widget index is unset.

File: src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp [hunk 2]
```diff
@@ -1829,7 +1828,7 @@ namespace OpenLoco::Ui::Windows::Construction::Construction
 
         if (_constructionHover == 1)
         {
-            if (!Input::isToolActive(WindowType::construction, self.number) || _toolWidgetIndex != widx::construct)
+            if (!Input::isToolActive(WindowType::construction, self.number) || ToolManager::getToolWidgetIndex() != widx::construct)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp:1831]: This is potentially the source of https://github.com/OpenLoco/OpenLoco/issues/1281. I want to do a few little tests this was incorrectly accessing the toolwidgetindex before.

Comment (duncanspumpkin) [reply src/OpenLoco/src/Windows/Construction/ConstructionTab.cpp:1831]: Yeah i'm pretty sure this is the cause of #1281.

File: src/OpenLoco/CMakeLists.txt [hunk 3]
```diff
@@ -198,6 +198,7 @@ set(OLOCO_SOURCES
     "${CMAKE_CURRENT_SOURCE_DIR}/src/ScenarioObjective.cpp"
     "${CMAKE_CURRENT_SOURCE_DIR}/src/SceneManager.cpp"
     "${CMAKE_CURRENT_SOURCE_DIR}/src/Title.cpp"
+    "${CMAKE_CURRENT_SOURCE_DIR}/src/ToolManager.cpp"
```
Comment (duncanspumpkin) [inline src/OpenLoco/CMakeLists.txt:201]: yeah think this should go in UI folder.

**Comment by duncanspumpkin** [comment]
Fixes #1281

File: src/OpenLoco/src/ToolManager.cpp (hunk 1 above)
Comment (LeftofZen) [reply src/OpenLoco/src/ToolManager.cpp:45]: it was signed in some places I removed the loco_global, and unsigned in others - what should it be? from your comment it sounds like unsigned?

Comment (duncanspumpkin) [reply src/OpenLoco/src/ToolManager.cpp]: Make it unsigned and check all users to make sure they use it expecting that. There is at least one that was doing a < 0 check to look for the 0xFFFF state.

File: src/OpenLoco/src/Input/MouseInput.cpp [hunk 4]
```diff
@@ -16,6 +16,7 @@
 #include "Objects/ObjectManager.h"
 #include "Objects/RoadObject.h"
 #include "SceneManager.h"
+#include "ToolManager.h"
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Input/MouseInput.cpp:19]:
```suggestion
#include "Ui/ToolManager.h"
```

File: src/OpenLoco/src/ToolManager.cpp (hunk 1 above)
Comment (LeftofZen) [reply src/OpenLoco/src/ToolManager.cpp]: good spot, I am actually very unsure about this now. almost every method the widget index is passed into reads the widget index as int16_t. there is one method that even converts it to a uint8_t, and one that checks its bits with a `1 << widgetIndex`. Basically there are 5 or 6 different types its implicitly cast to around the place. i didn't see a < 0 but I did see a >= 0.
Then, in Window.h there is a typedef `using WidgetIndex_t = int8_t` which is different AGAIN! But on that one, perhaps widgetIndex and toolWidgetIndex are truly different types? But I wouldn't have expected that.

File: src/OpenLoco/src/Input/MouseInput.cpp (hunk 4 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Input/MouseInput.cpp]: No worries. I missed the other includes in this quick fix, though. Could you address the CI errors?

File: src/OpenLoco/src/ToolManager.cpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/ToolManager.cpp]: I've changed it back to a signed `int16_t` as our CI was complaining about a comparison in `MouseInput.cpp` always being true. Could've changed that comparison into checking against 0xFFFF, but making the index signed seems the cleaner solution.

File: src/OpenLoco/src/Input/MouseInput.cpp [hunk 5]
```diff
@@ -335,15 +328,15 @@ namespace OpenLoco::Input
 
             resetMapSelectionFlag(MapSelectionFlags::enable | MapSelectionFlags::enableConstruct | MapSelectionFlags::enableConstructionArrow | MapSelectionFlags::unk_03 | MapSelectionFlags::unk_04);
 
-            if (_toolWidgetIndex >= 0)
+            if (ToolManager::getToolWidgetIndex() >= 0)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Input/MouseInput.cpp:331]: This needs to change into a check for `!= 0xFFFF`

File: src/OpenLoco/src/Window.cpp [hunk 6]
```diff
@@ -1482,7 +1483,7 @@ namespace OpenLoco::Ui
         uint64_t tool_widget = 0;
         if (Input::isToolActive(this->type, this->number))
         {
-            tool_widget = 1ULL << addr<0x00523394, uint32_t>();
+            tool_widget = 1ULL << ToolManager::getToolWidgetIndex();
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Window.cpp:1486]: Should investigate if this was causing any other bug as well.

Comment (duncanspumpkin) [reply src/OpenLoco/src/Window.cpp:1486]: Seems this is just UB. Unlikely to be the source of an additional bug.

File: src/OpenLoco/src/Input/MouseInput.cpp (hunk 5 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Input/MouseInput.cpp:331]: I changed the type to int16_t, so this should be fine.

File: src/OpenLoco/src/Ui/ToolManager.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    Ui::WindowNumber_t getToolWindowNumber();
+    void setToolWindowNumber(Ui::WindowNumber_t toolWindowNumber);
+
+    // 0x00523392
+    Ui::WindowType getToolWindowType();
+    void setToolWindowType(Ui::WindowType toolWindowType);
+
+    // 0x00523393
+    Ui::CursorId getToolWindowCursor();
+    void setToolWindowCursor(Ui::CursorId toolWindowCursor);
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/OpenLoco/src/Ui/ToolManager.h:21]: yeah probably a good idea.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**